File: MIDTERM_LAB/DataLoader.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class DataLoader {

    public static File promptForFile(Scanner scanner) {
        File file;
        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = scanner.nextLine().trim();

            if (path.startsWith("\"") && path.endsWith("\""))
                path = path.substring(1, path.length() - 1);

            file = new File(path);

            if (file.exists() && file.isFile()) {
                if (!path.toLowerCase().endsWith(".csv")) {
                    System.out.println("Invalid file path. Please try again.");
                    continue;
                }
                System.out.println("File found. Processing...");
                break;
            } else {
                System.out.println("Invalid file path. Please try again.");
            }
        }
        return file;
    }

    public static List<VideoGame> loadGames(File file) throws IOException {
        List<VideoGame> games = new ArrayList<>();

        BufferedReader br = new BufferedReader(
            new InputStreamReader(new FileInputStream(file), "UTF-8"));

        String line = br.readLine(); // skip header
        int skipped = 0;

        while ((line = br.readLine()) != null) {
            if (line.trim().isEmpty()) continue;
            VideoGame g = VideoGame.fromCSVLine(line);
            if (g != null) {
                games.add(g);
            } else {
                skipped++;
            }
        }
        br.close();

        System.out.println("Records loaded : " + games.size());
        if (skipped > 0)
            System.out.println("Rows skipped   : " + skipped);

        return games;
    }
}

File: MIDTERM_LAB/JS/DataLoader.js
// DataLoader.js
// Handles file path validation and CSV loading

const fs       = require('fs');
const readline = require('readline');
const VideoGame = require('./VideoGame');

function promptForFile(rl) {
    return new Promise((resolve) => {
        function ask() {
            rl.question('Enter dataset file path: ', function(path) {
                // Remove surrounding quotes if any
                path = path.trim().replace(/^"|"$/g, '');

                if (fs.existsSync(path) && fs.statSync(path).isFile()) {
                    if (!path.toLowerCase().endsWith('.csv')) {
                        console.log('Invalid file path. Please try again.');
                        ask();
                        return;
                    }
                    console.log('File found. Processing...');
                    resolve(path);
                } else {
                    console.log('Invalid file path. Please try again.');
                    ask();
                }
            });
        }
        ask();
    });
}

function loadGames(filePath) {
    try {
        const content = fs.readFileSync(filePath, 'utf8');
        const lines   = content.split('\n');

        const games   = [];
        let skipped   = 0;

        // Skip header (first line)
        for (let i = 1; i < lines.length; i++) {
            const line = lines[i].trim();
            if (!line) continue;

            const game = VideoGame.fromCSVLine(line);
            if (game) {
                games.push(game);
            } else {
                skipped++;
            }
        }

        console.log('Records loaded : ' + games.length);
        if (skipped > 0)
            console.log('Rows skipped   : ' + skipped);

        return games;

    } catch (e) {
        throw new Error('Failed to read file: ' + e.message);
    }
}

module.exports = { promptForFile, loadGames };